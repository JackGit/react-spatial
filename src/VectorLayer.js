import OLVectorLayer from 'ol/layer/Vector';
import VectorSource from 'ol/source/Vector';

import Layer from './Layer';

export default class VectorLayer extends Layer {
  /**
   * Vector layer.
   * @param {string} name The layer's name
   * @param {Object} [options] Layer options
   * @param {Array.<ol.Feature>} [options.features] Layer features
   * @param {ol.style.Style} [options.style] Layer style
   */
  constructor(options) {
    const {
      style, features, source, declutter,
    } = options;
    const olLayer = new OLVectorLayer({
      declutter,
      style,
      source: source || new VectorSource({
        features: features || [],
      }),
    });

    super({ ...options, olLayer });

    /**
     * List of features.
     * @type {Array.<ol.Feature>}
     */
  }

  /**
   * Add features to the layer.
   * @p@param {Array.<ol.Feature>} features List of feature.
   */
  addFeatures(features) {
    this.olLayer.getSource().addFeatures(features);
  }

  getFeatures() {
    return this.olLayer.getSource().getFeatures();
  }

  clear() {
    this.olLayer.getSource().clear();
  }
}
