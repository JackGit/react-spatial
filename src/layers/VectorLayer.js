import { buffer } from 'ol/extent';
import Layer from '../Layer';

/**
 * A class representing vector layer to display on BasicMap
 * @class
 */
class VectorLayer extends Layer {
  constructor(options = {}) {
    super(options);

    // Array of click callbacks
    this.clickCallbacks = [];

    // Add click callback
    if (options.onClick) {
      this.onClick(options.onClick);
    }
  }

  /**
   * Listens to click events on the layer.
   * @param {function} callback Callback function, called with the clicked
   *   features (https://openlayers.org/en/latest/apidoc/module-ol_Feature.html),
   *   the layer instance and the click coordinate.
   */
  onClick(callback) {
    if (typeof callback === 'function') {
      this.clickCallbacks.push(callback);
    } else {
      throw new Error('callback must be of type function.');
    }
  }

  /**
   * Request feature information for a given coordinate.
   * @param {ol.Coordinate} coordinate Coordinate to request the information at.
   * @returns {Promise<Object>} Promise with features, layer and coordinate
   *  or null if no feature was hit.
   * eslint-disable-next-line class-methods-use-this
   */
  getFeatureInfoAtCoordinate(coordinate) {
    const res = this.map.getView().getResolution();
    const extent = buffer([...coordinate, ...coordinate], 5 * res);
    const features = this.olLayer.getSource().getFeaturesInExtent(extent);

    return Promise.resolve({
      features,
      layer: this,
      coordinate,
    });
  }

  /**
   * Initialize the layer and listen to feature clicks.
   * @inheritDoc
   */
  init(map) {
    super.init(map);
    this.map = map;

    // Listen to click events
    this.map.on('singleclick', e => {
      if (!this.clickCallbacks.length) {
        return;
      }

      this.getFeatureInfoAtCoordinate(e.coordinate)
        .then(features => {
          this.clickCallbacks.forEach(c => c(features, this, e));
        })
        .catch(() => {
          this.clickCallbacks.forEach(c => c([], this, e));
        });
    });
  }
}

export default VectorLayer;
