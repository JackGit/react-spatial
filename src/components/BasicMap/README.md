#

This demonstrates the use of BasicMap.

```jsx
import React from 'react';
import VectorLayer from 'ol/layer/Vector';
import VectorSource from 'ol/source/Vector';
import GML from 'ol/format/GML';
import { bbox } from 'ol/loadingstrategy';
import BasicMap from 'react-spatial/components/BasicMap';
import ConfigReader from 'react-spatial/ConfigReader';
import TokenLayer from 'react-spatial/layers/TokenLayer';


const layers = ConfigReader.readConfig([{
  name: 'OSM Baselayer',
  visible: true,
  data: {
    type: 'xyz',
    url: 'https://c.tile.openstreetmap.org/{z}/{x}/{y}.png',
  },
}]);

layers.push(new TokenLayer({
  key:'tokenLayer',
  username: '',
  password: '',
  tokenUrl:'https://map.geo.fr.ch/arcgis/tokens/?request=gettoken',
  expiration: 1, // in minutes
  olLayer: new VectorLayer({
    source: new VectorSource({
      format: new GML(),
      loader(extent, resolution, projection) {
        if (!this.token) {
          return;
        }
        const proj = projection.getCode();
        const url = `https://map.geo.fr.ch/arcgis/services/SIPO/SIPO_Liegenschaften/MapServer/WFSServer?service=WFS&version=1.1.0&request=GetFeature&typename=RealEstate&srsname=${proj}&bbox=${extent.join(
          ',',
        )},${proj}&token=${(this && this.token) || ''}`;

        fetch(url)
          .then((response) => {
            return response.text();
          })
          .then((data) => {
            if (!data) {
              throw new Error('Empty data');
            }
            this.addFeatures(
              this.getFormat().readFeatures(data),
            );
          })
          .catch(() => {
            this.removeLoadedExtent(extent);
          });
      },
      strategy: bbox,
    }),
  }),
  onTokenUpdate: (token, layer) => {
    const source = layer.olLayer.getSource();
    source.token = token;
    source.refresh();
  },
}));

<BasicMap layers={layers} extent={[782540.2374443528, 5896333.64697159, 789429.4575714802, 5899261.077620511]} tabIndex={0}/>;
```
