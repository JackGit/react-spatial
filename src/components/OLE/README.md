#

This demonstrates the use of OLE.

```jsx
import React from 'react';
import BasicMap from 'react-spatial/components/BasicMap';
import ConfigReader from 'react-spatial/ConfigReader';
import Popup from 'react-spatial/components/Popup';
import ResizeHandler from 'react-spatial/components/ResizeHandler';
import Layer from 'react-spatial/Layer';
import VectorLayer from 'react-spatial/VectorLayer';
import OLMap from 'ol/Map';
import Feature from 'ol/Feature';
import Point from 'ol/geom/Point';
import VectorSource from 'ol/source/Vector';
import TileLayer from 'ol/layer/Tile';
import TileGrid from 'ol/tilegrid/TileGrid';
import TileImageSource from 'ol/source/TileImage';
import { getCenter } from 'ol/extent';
import Style from 'ol/style/Style';
import Circle from 'ol/style/Circle';
import Fill from 'ol/style/Fill';
import OLE from 'react-spatial/components/OLE';
import OSM, {ATTRIBUTION} from 'ol/source/OSM.js';
import 'ol/ol.css';

class OLEExample extends React.Component {
  constructor(props) {
    super(props);

    this.map = new OLMap();

    const feat = new Feature(new Point(map.getView().getCenter()));
    feat.setStyle(new Style({
      image: new Circle({
        radius: 10,
        fill: new Fill({
          color: '#ff0000',
        }),
      }),
    }));

    this.layers = [
      new Layer({
        olLayer:new TileLayer({
          source: new OSM()
        })
      }),
      new VectorLayer({
        source: new VectorSource({
          features: [feat]
        }),
      })
    ];
  }

  render() {
    return (
      <div className="tm-ole-example">
        <BasicMap
          map={this.map}
          layers={this.layers}
        />
        <OLE
          map={this.map}
          layer={this.layers[1]}
        />
      </div>
    );
  }
}

<OLEExample />;
```
