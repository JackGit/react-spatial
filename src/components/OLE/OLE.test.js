import 'jest-canvas-mock';
import React from 'react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';
import VectorLayer from 'react-spatial/VectorLayer';
import VectorSource from 'ol/source/Vector';
import OLMap from 'ol/Map';
import OLView from 'ol/View';
import BasicMap from 'react-spatial/components/BasicMap';
import OLE from './OLE';

configure({ adapter: new Adapter() });
const mockStore = configureStore();
const initialState = {
  dispatchSetCenter: () => {},
  dispatchSetZoom: () => {},
  dispatchSetResolution: () => {},
  dispatchSetLayers: () => {},

  center: [922747.8054581558, 5911639.7675754195],
  layers: [],
  extent: undefined,
  resolution: undefined,
  zoom: 9,
};

const olView = new OLView();
const map = new OLMap({ view: olView });
const vectorLayer = new VectorLayer({
  source: new VectorSource({
    features: [],
  }),
});
const layers = [vectorLayer];

describe('OLE', () => {
  test('matche snapshots', () => {
    const store = mockStore(initialState);
    window.ole = 'fgjkfgio';
    const component = renderer.create(
      <Provider store={store}>
        <BasicMap map={map} layers={layers} zoom={0} />
        <OLE map={map} layer={vectorLayer} />
      </Provider>,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
