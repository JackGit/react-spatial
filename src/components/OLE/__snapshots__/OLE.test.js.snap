// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OLE matche snapshots 1`] = `
<div
  className="tm-map"
  role="menu"
  tabIndex={0}
/>
`;
